src/services/route.rs

```diff
@@ -0,0 +1,257 @@
+            fragment: url.fragment().map(str::to_string)
+        }
+    }
+}
+
+
+impl RouteService {
+
+    /// Creates a new route service
+    pub fn new() -> RouteService {
+        RouteService {
+            history: window().history(),
+            location: window().location().unwrap(),
+            event_listener: None,
+            callback: None
+        }
+    }
+
+
+    /// Creates the callback used in the main routing logic.
+    ///
+    /// The callback takes a string, parses it into a url, and then uses the result of that
+    /// to create a message that the component will use to update itself with.
+    // TODO I would like to include this in register_router
+    pub fn create_routing_callback<COMP, CTX>(context: &mut Env<CTX, COMP>) -> Callback<RouteResult>
+        where
+            COMP: Component<CTX>,
+            COMP::Msg: From<RouteResult>,
+            CTX: 'static
+    {
+        return context.send_back(|route_result: RouteResult| {
+            println!("Callback path changed {:?}", route_result); // Remove me
+            COMP::Msg::from(route_result)
+        })
+    }
+
+    /// Will return the current route info based on the location API.
+    // TODO this should probably return a RouteResult and avoid expecting
+    pub fn get_current_route_info(&mut self) -> RouteInfo {
+        // If the location api errors, recover by redirecting to a valid address
+        let href = self.get_location().expect("Couldn't get href from location Api");
+        let url = Url::parse(&href).expect("The href returned from the location api should always be parsable.");
+        RouteInfo::from(url)
+    }
+
+    /// Registers the router.
+    /// There can only be one router.
+    /// The component in which it is set up will be the source from which routing logic emanates.
+    pub fn register_router(&mut self, callback: Callback<RouteResult>)
+    {
+        if let Some(_) = self.event_listener {
+            panic!("You cannot register two separate routers.");
+        }
+
+        // Hold on to the callback so it can be used to update the main router component
+        // when a user clicks a link, independent of the event listener.
+        self.callback = Some(callback.clone());
+
+        // Set the event listener to listen for the history's pop state events and call the callback when that occurs
+        self.event_listener = Some( window().add_event_listener(move |event: PopStateEvent| {
+            let state_value: Value = event.state();
+
+            if let Ok(state) = String::try_from(state_value) {
+                callback.emit(RouteInfo::parse(&state))
+            } else {
+                eprintln!("Nothing farther back in history, not calling routing callback.");
+            }
+        }));
+    }
+
+
+    /// Sets the route via the history api.
+    /// This does not by itself make any changes to Yew's state.
+    fn set_route(&mut self, route_info: RouteInfo) -> RouteDidChange {
```

Comment:
So this is where things get interesting.

The javascript form of this [`history.pushState`](https://developer.mozilla.org/en-US/docs/Web/API/History_API#The_pushState()_method) which accepts:
- the `state` object (arbitrary javascript value)
- a "title", not currently used
- a URL.

I think we should do the same. Requiring someone to construct `route_info` is unergonomic IMO.

Reply:
The `set_route` method is actually the main point of difference between our APIs, and I am convinced mine is better.

In my implementation it returns `COMP::Msg`. The reason it does this is to allow the _user_ to decide whether to propagate the "route change event". For instance, if you did `set_route` in response to a button click you could _choose_ whether to send the resulting message or send a different one. That is very intuitive IMO.

To do this all it has to do is call the wrapped `Fn` in the context, which we need #188 for.

Reply:
To address your first comment regarding the ergonomics, yes, creating RouteInfo structs is a burden placed upon the crate user, but the alternative is to pass strings around, which can lead to bugs at runtime. For example, the function responsible for mapping route strings to component properties could recieve a string that wasn't correctly formatted because the developer concatenated `"/parent/"` and `/child` together to create `/parent//child` and tried to use it for routing, when their mapping function expects `/parent/child`. By forcing users to parse their strings into RouteInfo, these errors can be prevented.

In the spirit of Rust, I'm trying to make what could otherwise be a runtime error into a compile time error.

Currently I have a pattern like: 
```rust
...
Route::DogForum => RouteInfo::parse("/dog").unwrap(),
...
```

which, yes, does throw a runtime error.

I intend to move this to something like:
```rust
Route::DogForum => route_info!("/dog"),
```
which would invoke the parse function at compile time, and stop compilation if the parser couldn't convert it to a RouteInfo.

It would also look like the following for when the user wants to switch routes.
```rust
context.routing.call_link(route_info!("/parent/child"));
```

If you wanted to ensure that you only call currently routable links and defend against the possibility of the route strings to specific components changing and all "links" to them no longer referring to that component:
```rust
context.routing.call_link(component_1::Route::Parent.into() + component_2::Route::Child.into());
```

or in the case of wanting to attach stateful information:
```rust
let route_info = route_info!("parent/child").append(some_user_specified_value.to_string());
context.routing.call_link(route_info);
```

I think this extra ceremony for creating navigable information provides safety benefits that outweigh the concerns about